### PR TITLE
feat: current action persistence on abrupt shutdowns

### PR DESCRIPTION
**Pull request by de-sh**
Closes #

### Changes

1. Uplink console's HTTP API now has two endpoints:
    - `/logs` : accepts a raw-string filter which is used to reload uplink's logger, i.e. tracing_subscriber
    - `/shutdown`: Triggers bridge to persist current action with a savefile, before leading to uplink shutting down entirely.

    NOTE: action's timeout works like a stopwatch and will only be paused on shutdown, i.e, if action started with timeout of 300s, shutdown was triggered on the clock having 150s left, on uplink restart, there will be 150s till action is deemed to have timedout.

2. Uplink commands over `console` is the new terminology replacing `apis` in config:
    ```
    - [apis]
    + [console]
    ```
3. Shutdown can also be triggered with POSIX IPC signals triggered on `Ctrl+C`, `reboot` or `shutdown`.

### Why?
Allows the user's application to trigger a shutdown in uplink while ensuring persistence of current action information. This is necessary to support OTA updates that might require uplink to be shutdown between action being started and action completion.

### Trials Performed

1. start uplink connected to port 5555 by an application that handles the "lock" action, and takes a long time to complete. i.e. following config
```
[tcpapps.1]
port = 5555
actions = [{ name = "lock", timeout = 300 }]

[console]
enabled = true
port = 3333
```
```
  2023-04-27T15:27:22.645246Z  INFO uplink::base::mqtt: Action = Action { device_id: None, action_id: "345", kind: "process", name: "lock", payload: "{}" }

  2023-04-27T15:27:22.645363Z  INFO uplink::base::bridge: Received action: Action { device_id: None, action_id: "345", kind: "process", name: "lock", payload: "{}" }

  2023-04-27T15:27:22.645414Z  INFO uplink::base::bridge: Action response = ActionResponse { action_id: "345", device_id: None, sequence: 0, timestamp: 1682609242645, state: "Received", progress: 0, errors: [], done_response: None }
...
```
3. From elsewhere, trigger a shutdown a second or two after above logs
```
#Terminal2:
 curl -X POST http://localhost:3333/shutdown
...
#Terminal1:
  2023-04-27T15:39:34.944052Z  INFO uplink::apis: Shutting down uplink

  2023-04-27T15:39:34.944181Z  INFO uplink::base::bridge: Storing current action in persistence; path: /tmp/uplink/current_action
```
5. Restart uplink, without action handler, hence no action completion response, triggering timeout
```
  2023-04-27T15:51:25.615515Z  INFO uplink::base::bridge: Loading saved action from persistence; action_id: 345
...
  2023-04-27T15:55:46.424650Z ERROR uplink::base::bridge: Timeout waiting for action response. Action ID = 345
```

### CODE REVIEW

**Comment by amokfa** [comment]
This won't save current action on device reboot. We can use this https://rust-cli.github.io/book/in-depth/signals.html

**Comment by de-sh** [comment]
Experimented on this here: https://github.com/bytebeamio/uplink/compare/abrupt-shutdown...signals?expand=1

File: uplink/src/base/serializer/mod.rs [hunk 1]
```diff
@@ -134,21 +136,23 @@ struct StorageHandler {
...
-            let mut storage = Storage::new(config.persistence.max_file_size);
-            if let Some(persistence) = &config.persistence.disk {
-                let mut path = persistence.path.clone();
+        for (stream_name, stream_config) in config.streams.iter() {
+            let mut storage = Storage::new(stream_config.persistence.max_file_size);
+            if stream_config.persistence.max_file_count > 0 {
+                let mut path = config.persistence_path.clone();
                 path.push(stream_name);
 
+                std::fs::create_dir_all(&path).map_err(|_| {
```
Comment (tekjar) [inline uplink/src/base/serializer/mod.rs:145]: Check if uplink is exiting with good error message during this case

Comment (tekjar) [reply uplink/src/base/serializer/mod.rs:145]: Create `negative cases` section in QA doc and add

- Unable to access persistence path
- Unable to create a dir
- Config file not available 
- Auth file not available
- Config fiile/Auth file is not valid json or toml

**Review by tekjar — CHANGES_REQUESTED** [review]

**Pull request merged**